File: backend/src/main/java/by/fpmibsu/findafriend/controller/commands/usershelter/AddUserToShelterHandler.java
package by.fpmibsu.findafriend.controller.commands.usershelter;

import by.fpmibsu.findafriend.application.mediatr.RequestHandler;
import by.fpmibsu.findafriend.dataaccesslayer.user.UserDao;
import by.fpmibsu.findafriend.dataaccesslayer.usershelter.UserShelterDao;
import by.fpmibsu.findafriend.entity.User;

public class AddUserToShelterHandler extends RequestHandler<Void, AddUserToShelterCommand> {
    private final UserDao userDao;
    private final UserShelterDao userShelterDao;

    public AddUserToShelterHandler(UserDao userDao, UserShelterDao userShelterDao) {
        this.userDao = userDao;
        this.userShelterDao = userShelterDao;
    }

    @Override
    public Void handle(AddUserToShelterCommand request) throws Exception {
        var user = userDao.getEntityById(request.userId);
        if (user == null) {
            return null;
        }
        if (User.Role.USER.equals(user.getRole())) {
            userDao.updateRole(user.getId(), User.Role.SHELTER_ADMINISTRATOR);
        }
        userShelterDao.removeUser(user.getId());
        userShelterDao.add(request.shelterId, request.userId);
        return null;
    }
}

File: BackendNET/Backend/Commands/UserShelter/AddUserToShelter/AddUserToShelterHandler.cs
using Backend.Entities;
using Backend.Repository;
using MediatR;

namespace Backend.Commands.UserShelter.AddUserToShelter;

public class AddUserToShelterHandler : IRequestHandler<AddUserToShelterCommand, Unit>
{
    private readonly IUserShelterRepository _userShelterRepository;
    private readonly IUserRepository _userRepository;

    public AddUserToShelterHandler(IUserShelterRepository userShelterRepository, IUserRepository userRepository)
    {
        _userShelterRepository = userShelterRepository;
        _userRepository = userRepository;
    }

    public async Task<Unit> Handle(AddUserToShelterCommand request, CancellationToken cancellationToken)
    {
        var user = await _userRepository.GetByIdAsync(request.UserId, cancellationToken);
        if (user is null)
        {
            return default;
        }

        if (user.Role == User.UserRole.User)
        {
            user.Role = User.UserRole.ShelterAdministrator;
            await _userRepository.UpdateAsync(user, cancellationToken);
        }

        await _userShelterRepository.RemoveByUserIdAsync(request.UserId, cancellationToken);
        var obj = new Entities.UserShelter { UserId = request.UserId, ShelterId = request.ShelterId };
        await _userShelterRepository.InsertAsync(obj, cancellationToken);
        return default;
    }
}